### PR TITLE
feat(channels/telegram): per-message reply modality override

### PR DESCRIPTION
**Pull request by jankovandeventer**
## Summary

Lets the user override reply modality on a per-message basis, in either direction:

- Voice in + *\"respond in text\"* → text reply
- Text in + *\"send a voice note\"* → voice reply (if TTS supported)

Previously the reply modality was hard-wired to the input modality (`willReplyWithVoice = isVoice && ttsSupported(...)`).

## How it works

- New `replyModalityOverride(text)` in `src/lib/audio.ts`. Conservative regex parser:
  - Matches near-leading directives: `\"reply with text\"`, `\"as voice\"`, `\"send a voice note\"`, `\"no voice\"`, etc.
  - **Negation short-circuit**: `\"do not reply with voice\"` → text.
  - **Later-wins**: if a message has competing directives, the last one wins (mid-sentence corrections work).
  - **Case-insensitive**.
  - **No phrase stripping** — the model can ignore the directive in its response if appropriate.
- Wired into `processChatMessage` **after STT** so transcripts of voice notes can also carry an override.
- Falls back to text gracefully if the user asks for voice but TTS is unsupported (existing TTS-failure path).
- `modalityOverride: \"text\" | \"voice\" | \"none\"` added to the completion log line for observability.

## What is NOT in this PR

- Per-chat sticky preference (a `/replytext` toggle). Easy follow-up.
- Stripping the directive phrase from the prompt fed to the model. Brittle and unnecessary — the model handles it fine.
- Localisation beyond English.

## Tests

- **25 new unit cases** in `tests/lib-audio.test.ts` covering:
  - 11 positive text overrides (\"reply with text\", \"text please\", \"no voice\", \"as text\", …)
  - 7 positive voice overrides (\"send a voice note\", \"voice reply\", \"reply with audio\", …)
  - 9 negative cases that should **NOT** trigger (e.g. \"compose a text message to John\", \"the chapter is text-heavy\", \"do not reply with voice\" should be text not voice)
  - 2 mid-sentence corrections (later-wins)
  - 1 case-insensitivity check
- **3 new e2e cases** in `tests/channels-telegram.test.ts`:
  - voice-in + override → text (TTS skipped)
  - text-in + override → voice (TTS fires)
  - text-in + voice override + no TTS → text (graceful fallback)

## Test plan

- [x] \`bun tsc --noEmit\` clean
- [x] \`bun test tests/lib-audio.test.ts tests/channels-telegram.test.ts\` → 117/117 pass
- [x] Full suite: 812 pass, 12 pre-existing failures unrelated to this change (\`cli-update.test.ts\` + \`cli-import-persona.test.ts\` — fail on \`main\` too)
- [ ] CI pipeline green
- [ ] One approval

## Docs

- `README.md`: new \"Per-message modality override\" sub-section under Voice replies.
- `AGENTS.md`: new pitfall entry explaining the routing rule and pointing future contributors at the regex + tests when adding phrasings.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/lib/audio.ts [hunk 1]
```diff
@@ -80,6 +80,89 @@ export function ttsSupported(config: Config): boolean {
...
+ *
+ * Deliberately conservative. Only matches explicit reply-form phrases
+ * ("respond with text", "as voice") and near-unmistakable shorthand
+ * ("voice note", "no voice"). Avoids false positives on bare nouns —
+ * "the chapter is text-heavy" must NOT trigger, nor "compose a text
+ * message to John". When the same message contains both a text and a
+ * voice directive (a user mid-correction), the later one wins.
+ */
+export function replyModalityOverride(
+  text: string | undefined,
```
Comment (andrewagrahamhodges) [inline src/lib/audio.ts:104]: 🔵 **Trivial**: This regex is well-constructed. The use of word boundaries and non-capturing groups for optional words like 'me a' ensures it's both flexible and conservative against false positives.

File: src/channels/telegram.ts [hunk 2]
```diff
@@ -972,8 +973,26 @@ async function processChatMessage(
...
+  //                   text-in → text-out)
+  //   user override — explicit per-message directive in the message
+  //                   text ("reply in text", "send a voice note") flips
+  //                   the wire format. Applies AFTER STT so the
+  //                   transcript is what gets inspected for voice-in.
+  //
+  // Voice is still capped by ttsSupported(): if the user asks for a
+  // voice reply but the provider can't do TTS, we degrade to text
+  // gracefully (same fallback as the original voice-in path).
+  const modalityOverride = replyModalityOverride(msg.text);
```
Comment (andrewagrahamhodges) [inline src/channels/telegram.ts:988]: 🔵 **Trivial**: Correct application of the priority: override wins, otherwise mirror input, but always capped by TTS availability. Clean and logical.

**Review by andrewagrahamhodges — APPROVED** [review]
## Code Review: feat(channels/telegram): per-message reply modality override

**Author:** @jankovandeventer | **Files:** 6 | **Verdict:** Approve

### Findings

| # | Severity | File | Line | Finding |
|---|----------|------|------|---------|
| 1 | 🔵 Trivial | src/lib/audio.ts | 104 | Regex for negation of voice |
| 2 | 🔵 Trivial | src/channels/telegram.ts | 988 | Logic for applying the modality override |

### What's Good
- Solid Regex design in `replyModalityOverride`: using word boundaries (`\b`) and non-capturing groups (`(?:...)`) correctly prevents false positives on bare nouns.
- Strategic placement of the override check *after* STT in the Telegram channel ensures that voice-in transcripts can also trigger a modality flip.
- Excellent test coverage with 25 unit cases and 3 integration scenarios, including the "later-wins" logic for user corrections.
- Graceful fallback for unsupported TTS providers is correctly maintained.
- Documentation updates in `README.md` and `AGENTS.md` are clear and helpful for both users and future developers.

<details>
<summary>📋 AI Fix Prompt (copy to your AI editor)</summary>

```
Review and fix the following issues found in PR #104 (feat(channels/telegram): per-message reply modality override):

1. [Trivial] All changes in src/lib/audio.ts and src/channels/telegram.ts are correct and well-implemented.

Repository: phantomyard/phantombot
Branch: feat/reply-modality-override
Base: main
```
</details>

**Review by kaieriksen — COMMENTED** [review]
## Kai's Review

This is a clean, well-executed PR. Janko clearly knows what he's doing.

### What I checked

- **Tests**: 25 unit + 3 e2e, all passing in CI. Coverage hits all three routing paths (voice→text, text→voice, text→voice-no-TTS), edge cases (later-wins, negation, case-insensitivity), and the negative cases that must NOT fire ("compose a text message to john", "text-heavy"). Solid.
- **Safety**: The regex is deliberately conservative — anchored on reply-verbs, word boundaries everywhere, no bare-noun matches. Negation short-circuits before positive patterns so "do not reply with voice" doesn't paradoxically trigger voice. The worst case is an unrecognised phrase → falls through to undefined → default mirror behaviour. No crash surface.
- **Regression risk**: Near zero. The function is additive (new export, new call site). The only changed lines in `telegram.ts` replace a one-liner with a three-liner that preserves the original fallback. Existing tests untouched.

### One observation (non-blocking)

In `replyModalityOverride`, the `NEGATION_VOICE` regex only covers negation of *voice* directives. The comment in the code acknowledges this is asymmetric — "no text" / "don't reply with text" isn't handled. That's fine for now, but if someone uses that phrasing while also having a voice directive elsewhere in the message, the later-wins logic could produce a surprising result. Worth a TODO or a follow-up issue if you want to close the gap later.

### Verdict

**Approve.** Safe to merge whenever you're ready.

File: src/lib/audio.ts [hunk 3]
```diff
@@ -80,6 +80,89 @@ export function ttsSupported(config: Config): boolean {
...
+  const t = text.toLowerCase();
+
+  // Negation short-circuit. "Do not reply with voice" literally CONTAINS
+  // the substring "reply with voice" which would otherwise hit the
+  // positive voice pattern below — so we check negations first and exit.
+  // Symmetric "no text / don't reply with text" isn't covered: it's an
+  // unlikely phrasing, and adding it just bloats the surface. Easy to
+  // bolt on if anyone asks.
+  const NEGATION_VOICE =
+    /\b(?:no|never|don't|do\s+not)\s+(?:use\s+|reply\s+(?:with|in)\s+|respond\s+(?:with|in)\s+|send\s+(?:me\s+|a\s+)*)?voice(?:\s*(?:note|message|reply|response))?\b/;
```
Comment (kaieriksen) [inline src/lib/audio.ts:116]: Nit (non-blocking): the negation pattern only covers voice→text. "Don't reply with text, use voice" would correctly hit the positive voice pattern (negation doesn't catch it), but a user with competing directives where text-negation isn't handled could see surprising later-wins results. The comment already acknowledges the asymmetry — just flagging for completeness.
